File: dotnet/src/Wata.Sdk/Errors/WataErrorCodes.cs
namespace Wata.Sdk.Errors;

/// <summary>
/// Семейство, к которому относится код ошибки WATA.
/// </summary>
public enum WataErrorFamily
{
    Unknown,
    PaymentLink,
    Crypto,
    Transaction,
    Refund,
    Order,
    Steam,
    Stars,
    TopUp,
    Voucher,
}

/// <summary>
/// Известные коды ошибок WATA из поля <c>error.code</c>.
/// </summary>
/// <remarks>
/// Список неполный по своей природе: платформа может добавить код без изменения
/// версии SDK. Поэтому это справочник констант, а не перечисление —
/// <see cref="WataApiException.Code"/> всегда содержит исходную строку,
/// даже незнакомую.
/// </remarks>
public static class WataErrorCodes
{
    // Платёжные ссылки.
    public const string LinkNotFound = "PL_1001";
    public const string LinkInvalid = "PL_1002";
    public const string LinkExpired = "PL_1003";

    // Шифрование карточных данных.
    public const string CryptoInvalid = "CRY_1001";

    // Возвраты. Транзакции: TRA_1001..TRA_1019 — валидация,
    // TRA_2001..TRA_2999 — отказы шлюза и эмитента.
    public const string RefundInvalidAmount = "TRA_1101";
    public const string RefundInsufficientFunds = "TRA_1102";
    public const string RefundPendingExists = "TRA_1103";

    /// <summary>
    /// Определяет семейство ошибки по префиксу кода. Позволяет обработать целую
    /// группу отказов, не перечисляя каждый код по отдельности.
    /// </summary>
    public static WataErrorFamily FamilyOf(string? code) => code switch
    {
        null or "" => WataErrorFamily.Unknown,
        _ when code.StartsWith("PL_", StringComparison.Ordinal) => WataErrorFamily.PaymentLink,
        _ when code.StartsWith("CRY_", StringComparison.Ordinal) => WataErrorFamily.Crypto,
        _ when code.StartsWith("TRA_11", StringComparison.Ordinal) => WataErrorFamily.Refund,
        _ when code.StartsWith("TRA_", StringComparison.Ordinal) => WataErrorFamily.Transaction,
        _ when code.StartsWith("ORD_", StringComparison.Ordinal) => WataErrorFamily.Order,
        _ when code.StartsWith("STM_", StringComparison.Ordinal) => WataErrorFamily.Steam,
        _ when code.StartsWith("STR_", StringComparison.Ordinal) => WataErrorFamily.Stars,
        _ when code.StartsWith("TPP_", StringComparison.Ordinal) => WataErrorFamily.TopUp,
        _ when code.StartsWith("VCR_", StringComparison.Ordinal) => WataErrorFamily.Voucher,
        _ => WataErrorFamily.Unknown,
    };
}

File: java/src/main/java/pro/wata/sdk/errors/WataErrorCodes.java
package pro.wata.sdk.errors;

/**
 * Известные коды ошибок WATA из поля {@code error.code}.
 *
 * <p>Список неполный по своей природе: платформа может добавить код без изменения
 * версии SDK. Поэтому это справочник констант, а не перечисление — код ошибки
 * всегда доступен как исходная строка, даже незнакомая.
 */
public final class WataErrorCodes {

    private WataErrorCodes() {
    }

    // Платёжные ссылки.
    public static final String LINK_NOT_FOUND = "PL_1001";
    public static final String LINK_INVALID = "PL_1002";
    public static final String LINK_EXPIRED = "PL_1003";

    // Шифрование карточных данных.
    public static final String CRYPTO_INVALID = "CRY_1001";

    // Возвраты. Транзакции: TRA_1001..TRA_1019 — валидация,
    // TRA_2001..TRA_2999 — отказы шлюза и эмитента.
    public static final String REFUND_INVALID_AMOUNT = "TRA_1101";
    public static final String REFUND_INSUFFICIENT_FUNDS = "TRA_1102";
    public static final String REFUND_PENDING_EXISTS = "TRA_1103";

    /** Семейство, к которому относится код ошибки WATA. */
    public enum Family {
        UNKNOWN,
        PAYMENT_LINK,
        CRYPTO,
        TRANSACTION,
        REFUND,
        ORDER,
        STEAM,
        STARS,
        TOPUP,
        VOUCHER
    }

    /**
     * Определяет семейство ошибки по префиксу кода. Позволяет обработать целую
     * группу отказов, не перечисляя каждый код по отдельности.
     *
     * @param code код из ответа WATA, может быть {@code null}
     * @return семейство ошибки либо {@link Family#UNKNOWN}
     */
    public static Family familyOf(String code) {
        if (code == null || code.isEmpty()) {
            return Family.UNKNOWN;
        }
        if (code.startsWith("PL_")) {
            return Family.PAYMENT_LINK;
        }
        if (code.startsWith("CRY_")) {
            return Family.CRYPTO;
        }
        if (code.startsWith("TRA_11")) {
            return Family.REFUND;
        }
        if (code.startsWith("TRA_")) {
            return Family.TRANSACTION;
        }
        if (code.startsWith("ORD_")) {
            return Family.ORDER;
        }
        if (code.startsWith("STM_")) {
            return Family.STEAM;
        }
        if (code.startsWith("STR_")) {
            return Family.STARS;
        }
        if (code.startsWith("TPP_")) {
            return Family.TOPUP;
        }
        if (code.startsWith("VCR_")) {
            return Family.VOUCHER;
        }
        return Family.UNKNOWN;
    }
}
